File: Week_5/bowling/lib/game.rb
require_relative './frame'
require_relative './tenth_frame'

class Game
  attr_reader :frames

  def initialize
    @frames = []
  end

  def calculate_current_score
    score = 0
    spare, strike, consecutive_strike = false

    @frames.each { |frame|
      score += frame.roll_1 if spare
      score += (frame.roll_1 + frame.roll_2) if strike
      score += (frame.roll_1) if consecutive_strike
      score += frame.roll_3 if frame.is_a?(TenthFrame)
      score += (frame.roll_1 + frame.roll_2)
      consecutive_strike = strike && frame.strike
      spare, strike = frame.spare, frame.strike
    }
    return score
  end


  def add_frame(roll_1, roll_2 = 0, roll_3 = 0)
    return "The game has finished" if @frames.length == 10
    return @frames << Frame.new(roll_1, roll_2) if @frames.length < 9
    return @frames << TenthFrame.new(roll_1, roll_2, roll_3) if @frames.length == 9
  end

end

File: Week_6/bowling/lib/game.js
const Frame = require('./frame');
const TenthFrame = require('./tenthframe');

class Game {

  constructor() {
    this.frames = []
  }

  addFrame(roll1, roll2, roll3) { 
    if (this.frames.length == 9) this.frames.push(new TenthFrame(roll1, roll2, roll3));
    if (this.frames.length < 9) this.frames.push(new Frame(roll1, roll2));
    else {return "No more throws"}
  }
 
  score() {
    let total = 0

    this.frames.map((fr,idx) => {
      const strike = fr.roll1 === 10 
      const spare = fr.roll1 != 10 && fr.roll1 + fr.roll2 === 10
      const nextFrame = this.frames[idx+1]
      const nextNextFrame = this.frames[idx+2]
      const tenth = this.frames.length === 10

      if(strike && nextNextFrame != null && nextFrame.roll1 === 10) 
        {fr.roll1 += (nextFrame.roll1 + nextNextFrame.roll1)} // consecutive strikes where this is both a next frame and nextnext frame

      if(strike && nextFrame != null && nextNextFrame === null && nextFrame === strike) 
        {fr.roll1 += (nextFrame.roll1 + nextFrame.roll2)} // consecutive strikes where there is no nextNextFrame

      if(fr.roll1 === 10 && nextFrame != null && nextFrame != strike) 
        {fr.roll1 += (nextFrame.roll1 + nextFrame.roll2)} // normal strikes

      if(spare && nextFrame != null) 
        {fr.roll1 += (nextFrame.roll1)}  // spares

      if(tenth && strike || spare) 
      {fr.roll1 += (fr.roll3 || 0)} // tenth round
      
      total += (fr.roll1 + fr.roll2);
    });
    return total;
  };
}

module.exports = Game